HARJA-1370-johtojahallintakorvaus-muutokset

jarnovayrynen opened the pull request:
Johto- ja hallintokorvausten muutokset: tietomalli, lomake, tallennus ja testit

jarnovayrynen committed:
Aloita johto- ja hallintokorvauksen muutoslomaketta

jarnovayrynen committed:
Aloita kulun tallennus JJH-muutoksessa

jarnovayrynen committed:
Siirrä migraationumeroa

jarnovayrynen committed:
Merge branch 'develop' of github.com:finnishtransportagency/harja into HARJA-1370-johtojahallintakorvaus-muutokset

jarnovayrynen committed:
Näytä MHU+ urakan sopimustyyppi

jarnovayrynen committed:
Tallenna johto- ja hallintokorvaus muutos kuluina

-Korjaa triggeriä päivittämään validi_aikana oikein
-Luo Kajaanin testiurakkaan toimenpideinstanssit

jarnovayrynen committed:
Merge branch 'develop' of github.com:finnishtransportagency/harja into HARJA-1370-johtojahallintakorvaus-muutokset

rckopteri commented on src/clj/harja/kyselyt/muutos_kyselyt.sql:
```diff
@@ -8,6 +8,14 @@ SELECT m.id,
+                                  JOIN mhu_muutos_kulu mmk ON (k.id = mmk.kulu AND m.id = mmk.muutos AND m.versio = mmk.versio)
+                                  JOIN kulu_kohdistus kk ON k.id = kk.kulu AND kk.tyyppi = 'jjh-muutos'
+                                  WHERE k.poistettu IS FALSE AND kk.poistettu IS FALSE
+                                   AND erapaiva BETWEEN (SELECT TO_DATE(:hoitokauden_alkuvuosi || '-10-01', 'YYYY-MM-DD')) AND
```
erapaiva kolumniviittaukseen voisi laittaa taulun, josta se tulee, eli mmk. kk. tai k.

rckopteri commented on src/clj/harja/kyselyt/muutos_kyselyt.sql:
```diff
@@ -74,15 +82,62 @@ VALUES
 
--- name: paivita-muutos!
+-- name: luo-muutos<!
+    -- name: luo-muutos<!
```
funktionimi kahdesti?

rckopteri commented on src/clj/harja/kyselyt/muutos_kyselyt.sql:
```diff
@@ -74,15 +82,62 @@ VALUES
+    (urakka, tyyppi, nimi, syy, kulu_kohdistus, luonnos, voimassa_alkaen, luoja, luotu)
+    VALUES
+    (:urakka, :tyyppi::MHU_MUUTOSTYYPPI, :nimi, :syy, :kulu_kohdistus, :luonnos, :voimassa_alkaen, :kayttaja, NOW())
+    RETURNING id, versio;
```
Ilman RETURNINGiä <! määreet varmistaa, että se palauttaa koko insertin

jarnovayrynen committed:
Katselmointikorjauksia

rckopteri commented on src/clj/harja/palvelin/palvelut/muutos/muutos_palvelu.clj:
```diff
@@ -95,16 +107,113 @@
+
+                  ;; tähän puuttuvien muutostyyppien lomakehaut...
+                  nil)
+        _ (assert (= 1 (count vastaus)) "Muutoksia täytyy palautua tasan yksi")
```
Mä vähän vierastan assertin käyttöä tuotantokoodissa. 
"_assertions are designed to be turned off in production…they are designed to catch logically impossible situations that can only occur if the program is wrong. they should not be used to detect invalid input (like a bad web request, bad user input, etc)._"
Täältä : https://clojureverse.org/t/pragmatic-programmer-should-we-all-use-assert/9106/6

jarnovayrynen replied:
poistin assertin, throwaan jos on tällainen virhetilanne

rckopteri commented on src/clj/harja/palvelin/palvelut/muutos/muutos_palvelu.clj:
```diff
@@ -95,16 +107,113 @@
+(defn- poista-vanhat-kulutiedot!
+  "Asettaa poistetuksi vanhat kulutiedot, jotta ne eivät näy käyttöliittymässä tai raporteissa."
+  ;; halutaan saada historiatieto talteen, tämä on siihen käytännöllinen tapa tekemättä valtavaa refaktorointia kulu tauluun (ja sille omaa historiataulua ja triggereitä)
+  [db user rivi]
```
userin vois nimetä käyttäjäksi

jarnovayrynen replied:
laitoin sen userin, [db user antaa 605 hittiä, [db kayttaja 104, niin aattelin että se user on tohon kohti vakiintunut :)

rckopteri commented on src/clj/harja/palvelin/palvelut/muutos/muutos_palvelu.clj:
```diff
@@ -95,16 +107,113 @@
+                                      (tpi-q/hae-urakan-toimenpideinstanssi-toimenpidekoodilla db
+                                        {:urakka (:id urakka)
+                                         :koodi (mhu/toimenpide-avain->toimenpide :mhu-johto)})))]
+    (doseq [rivi rivit
```
Saiskohan rivistä attribuutit esille myös näin
{:keys [pvm numero]}

jarnovayrynen replied:
Samoten tossa missä ehdotit rivin destrukturointia, siinä on muutama kenttä mitä parsitaan niistä muista mölleistä, kuten urakka ja user, niin jättäisin sen melkeen as is:
[rivi rivit
        :let [kulu {:urakka (:id urakka)
                    :erapaiva (:pvm rivi)
                    :numero (:numero rivi)
                    :kayttaja (:id user)
                    :lisatieto "Muutoksesta automaattisesti luotu kulu"
                    :laskun_numero (:laskun-numero rivi)
                    :koontilaskun-kuukausi (kulut-domain/pvm->koontilaskun-kuukausi (:pvm rivi) (:alkupvm urakka))
                    :kokonaissumma (:tavoitehinnan-muutos rivi)}

rckopteri commented on src/clj/harja/palvelin/palvelut/muutos/muutos_palvelu.clj:
```diff
@@ -95,16 +107,113 @@
-      ;; TODO: muutoksen tallennus tähän
-      )))
+      (let [muutos-paluurivi (if (:id muutos)
+                               (muutos-kyselyt/paivita-muutos<!
```
Voisin melkein laittaa yksiriviseksi

rckopteri commented on src/cljs/harja/tiedot/urakka/muutos_tiedot.cljs:
```diff
@@ -74,6 +102,14 @@
       muutoksien-kayttoonoton-hoitokauden-alkuvuosi)))
 
+(defn alusta-tyyppikohtaisia-arvoja [tyyppi valittu-hoitokausi]
+  (case tyyppi
```
Tuleeko tähän case:en myöhemmin enemmän ehtoja? Yhtä varten esim if voisi olla helpompi lukea?

jarnovayrynen replied:
Joo, tähän tulee sitä myöten kun hommat etenee, sillä tein poikkeuksellisesti vähän etupeltoon futureproofiksi. Ehkä jo Samulin taskissa tulee

rckopteri commented on src/cljs/harja/tiedot/urakka/muutos_tiedot.cljs:
```diff
@@ -159,10 +221,23 @@
+  HaeMuutoksenTiedot
+  (process-event [{muutos :muutos} app]
+    (let [valittu-hoitokausi (:valittu-hoitokausi app)]
+      (if (:id muutos)
```
Nämä katselmointityökalut ei ole aina ihan priimaa, mutta onko tässä iffissä kaksi palautustyyppiä , siis truelle ja falselle omansa? Ja palauttaahan tämä eventti aina app -staten lopuksi?

jarnovayrynen replied:
Tämä muuttui vielä hieman, huomioi mahdollliset pakollisten tietojen puutteet. Palauttaa aina app-staten.

rckopteri commented on src/cljs/harja/tiedot/urakka/muutos_tiedot.cljs:
```diff
@@ -108,20 +146,44 @@
-         :epaonnistui ->TallennaMuutosEpaonnistui})))
+        {:onnistui ->HaeUrakanMuutostiedotOnnistui         ;; voidaan käyttää samaa eventtiä, koska haetaan uudet muutostiedot tallennuksen jälkeen
+         :epaonnistui ->TallennaMuutosEpaonnistui})
+      app))
```
Yleisesti olisi hyvä, että aina kun käynnistetään bäkkäriin tallennus, niin laitettaisiin lippu -> :tallennus-kesken? true päälle ja sen perusteella laitetaan disablediksi kaikki nappulat ja vaikka jotain latausindikaattoreita.

jarnovayrynen replied:
tehty

jarnovayrynen committed:
Lisää testi JJH muutoksen tallentamiseen

jarnovayrynen committed:
Katselmointikommenttien pohjalta parannuksia

jarnovayrynen committed:
Korjaa testejä ja lisää testidataan JJH-muutos

jarnovayrynen committed:
Lisää testi yksittäisen muutoksen hakeminen

jarnovayrynen committed:
Lisää virheen palautus jos lomakkeelle palautuisi enemmän kuin yksi virhe

jarnovayrynen committed:
Siirrä migraationumeroa

jarnovayrynen committed:
Merge branch 'develop' of github.com:finnishtransportagency/harja into HARJA-1370-johtojahallintakorvaus-muutokset

jarnovayrynen committed:
Korjaa muutoksen päivitys

Kun muutosta päivitetään, täytyy käyttää avainsanaa UPDATE ONLY mhu_muutos, muuten update kohdistuu myös mhu_muutos_historia tauluun, koska se on tehty INHERIT:n-avulla.

Lisää tallennus-kesken? frontille, estää duplikaattitallennuksen.

jarnovayrynen committed:
Paranna sopimustyypin valintalaatikon tyyliä

jarnovayrynen committed:
Näytä käyttäjälle jos pakollisia kenttiä puuttuu

jarnovayrynen committed:
Korjaa testin odotettu tulos

Testidataan lisätty kulu inkrementoi odotettua id:tä

jarnovayrynen committed:
Validoi: 2025 tai myöh. JJH-muutos on oltava vähennys

Tuo mahdollinen virhe UI:lle asti ja näytä käyttäjälle.

jarnovayrynen committed:
Dokumentoi paremmin :paasta-virhe-lapi?

jarnovayrynen committed:
Rivitä sql-insert totutummin
